control-plane/agents/core/src/core/reconciler/poller.rs

```diff
@@ -0,0 +1,81 @@
+use crate::core::{
+    reconciler::{persistent_store::PersistentStoreReconciler, volume},
+    registry::Registry,
+    task_poller::{PollContext, PollEvent, PollResult, PollerState, TaskPoller},
+};
+
+/// Reconciliation worker that polls all reconciliation loops
+/// The loops are polled one at a time to avoid any potential contention
+/// and also hopefully making the logging clearer
+pub(super) struct ReconcilerWorker {
+    poll_targets: Vec<Box<dyn TaskPoller>>,
+    event_channel: tokio::sync::mpsc::Receiver<PollEvent>,
+    shutdown_channel: tokio::sync::mpsc::Receiver<()>,
+}
+impl std::fmt::Debug for ReconcilerWorker {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Reconciler").finish()
+    }
+}
+impl ReconcilerWorker {
+    /// Create a new `Self` with the provided communication channels
+    pub(super) fn new(
+        event_channel: tokio::sync::mpsc::Receiver<PollEvent>,
+        shutdown_channel: tokio::sync::mpsc::Receiver<()>,
+    ) -> Self {
+        Self {
+            poll_targets: vec![
+                Box::new(volume::VolumeReconciler::new()),
+                Box::new(PersistentStoreReconciler::new()),
+            ],
+            event_channel,
+            shutdown_channel,
+        }
+    }
+}
+
+impl ReconcilerWorker {
+    /// Start polling the registered reconciliation loops
+    /// The polling will continue until we receive the shutdown signal
+    #[tracing::instrument(skip(registry))]
+    pub(super) async fn poller(mut self, registry: Registry) {
+        // kick-off the first run
+        let mut event = PollEvent::TimedRun;
+        loop {
+            let result = match &event {
+                PollEvent::Shutdown => {
+                    tracing::warn!("Shutting down... (reconcilers will NOT be polled again)");
+                    return;
+                }
+                PollEvent::TimedRun | PollEvent::Triggered(_) => {
+                    self.poller_work(PollContext::from(&event, &registry)).await
+                }
+            };
+
+            event = tokio::select! {
+                _shutdown = self.shutdown_channel.recv() => {
+                    PollEvent::Shutdown
+                },
+                event = self.event_channel.recv() => {
+                    event.unwrap_or(PollEvent::Shutdown)
```

Comment:
Yes, a None means the channel has been dropped. It shouldn't happen as the shutdown should has priority in the select! but in case the select checks the shutdown first, and we're not dropped yet, and then check the event and we get dropped at that point.